clinvar_submissions_v1.1.0

RSWilson1 opened the pull request:
Re-write to add functionality to use clarity_extract or samples_file csv.
Updated SQL alchemy to fix issue (https://github.com/eastgenomics/clinvar_submissions/issues/14).
Added no-retry option (https://github.com/eastgenomics/clinvar_submissions/issues/20).
Added logic to avoid re-submission of duplicate variants (https://github.com/eastgenomics/clinvar_submissions/issues/18).
Added fix for https://github.com/eastgenomics/clinvar_submissions/issues/19.
Updated docstring (https://github.com/eastgenomics/clinvar_submissions/issues/17).
Changed formatting in line with black.
Changed print typo: https://github.com/eastgenomics/clinvar_submissions/issues/16.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23)
<!-- Reviewable:end -->


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Path-based clarity-extract mode with optional path validation, duplicate/missing detection, timestamped CSV outputs, organisation selection, and an updated mutually exclusive run-mode CLI with renamed flags; process map image version bumped.

- **New Utilities**
  - File-existence validation, DNAnexus authentication helper and output-directory validation; improved path handling and existence filtering.

- **Documentation**
  - README and examples updated to reflect new flags, run-modes and organisation option.

- **Tests**
  - New test suite covering clarity-extract parsing, path resolution, duplicates and edge cases.

- **Chores**
  - CI workflow and Python version updated; project metadata and dependencies added; repository CSV-ignore pattern adjusted.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

RSWilson1 committed:
Added script from RD-requests DI-1743 as clinvar_extract_handler.py.
Called this and added args for running in this mode with --clarity_extract.
Updated test config with new variables.

RSWilson1 committed:
fixed duplicate removal functgion and switched to regex
removed prints
changed raising errors from pandas csv reader

RSWilson1 committed:
Add unit tests for clarity_extract_handler functions

RSWilson1 committed:
change to rd_assays for future changes

RSWilson1 committed:
Added check that file exists on file system and removes files which don't

RSWilson1 committed:
Changed clarity handling to deal with new clarity extract to get assay from filename.
Now:
- Handled assays together for rare disease and in the future cancer
- Checks if R-code in filename is present in clarity extract for row.
- Rm's any rows where R-code doesn't match expected.
- Added prints for testing purposes

RSWilson1 committed:
Linting with ruff and added new function for checking if files exist

RSWilson1 committed:
Moved check_files_exist_and_exclude function to utils

RSWilson1 committed:
removed in_place to not return None

RSWilson1 committed:
removed extra rm for CNVs as not needed.
added rm for mosaics
updated function name
rm'd in_place =true to not return None

RSWilson1 committed:
Rm'd debug prints
changed to use pathlib
whitespace

RSWilson1 committed:
Added docstrings for each test

RSWilson1 committed:
rm'd future config update as broke test

RSWilson1 committed:
Added --use-paths to allow for running.
Added output of CSV with paths if not using --use-paths

RSWilson1 committed:
Add output for inconsistent files

RSWilson1 committed:
standardise args to use '_'

RSWilson1 committed:
Update readme with all standardised args and example cmd.
change prints to match new args

RSWilson1 committed:
Updated to use path_column
Output missing files for both samples_files and clarity_extract

RSWilson1 committed:
Fix test to use list instead of str

RSWilson1 committed:
Fixed find_file_name and added tests

RSWilson1 committed:
removed headers from runtime error

RSWilson1 committed:
Fix check_files_and_exclude function to return 2 dfs under all conditions.

RSWilson1 committed:
rm'd duplicate tests

RSWilson1 commented on tests/test_clarity_extract_handler.py:
```diff
@@ -0,0 +1,187 @@
+    assert "Error reading clarity extract:" in str(exc.value)
+
+
+@patch("utils.clarity_extract_handler.dxpy.find_projects")
+def test_get_matching_projects(mock_find_projects):
+    """Test getting matching projects in normal scenario."""
+    mock_find_projects.return_value = [
+        {"id": "proj-1", "describe": {"name": "002_foo_CEN"}},
+        {"id": "proj-2", "describe": {"name": "002_bar_CEN"}},
+    ]
+    result = ceh.get_matching_projects("CEN")
+    assert result == [("proj-1", "002_foo_CEN"), ("proj-2", "002_bar_CEN")]
+
```
changed
<!-- Reviewable comment -Ob8eCAc8OAddBRbCkDr:b-n9tkhd -->
<!-- Sent from Reviewable.io -->

RSWilson1 commented on utils/clarity_extract_handler.py:
```diff
@@ -0,0 +1,423 @@
+    for record in all_records:
+        record["project_name"] = project_dict.get(record["project_id"], "Unknown")
+
+    # Create a DataFrame from the records
+    records_df = pd.DataFrame(all_records)
+    if records_df.empty:
+        print("No records found for the given sample IDs.")
+        return df
+    # Merge with the original df to retain additional columns
+    merged_df = pd.merge(df, records_df, on="sample_id", how="left")
+    print(f"Merged DataFrame shape: {merged_df.shape}")
+    print(merged_df.head())
```
changed
<!-- Reviewable comment -ObDelw23fCu8adPvWep:b-n9tkhd -->
<!-- Sent from Reviewable.io -->

RSWilson1 commented on utils/clarity_extract_handler.py:
```diff
@@ -0,0 +1,423 @@
+            return None
+        else:
+            return filenames[0]
+
```
changed
<!-- Reviewable comment -ObDemwGBtgxQAwFKx7Q:b-n9tkhd -->
<!-- Sent from Reviewable.io -->

RSWilson1 commented on utils/utils.py:
```diff
@@ -583,8 +564,12 @@ def submission_status_check(
+            "Status check failed:\n"
+            + str(headers)
+            + "\n"
+            + url
+            + "\n"
+            + response_content
         )
```
changed
<!-- Reviewable comment -Ob8q9AQ7LvNP99zn6Yv:b-n9tkhd -->
<!-- Sent from Reviewable.io -->

RSWilson1 commented on utils/utils.py:
```diff
@@ -618,10 +600,50 @@ def submission_status_check(
+
+    path = Path(path)  # ensure it’s a Path object
+
+    return path.exists()
+
+
+def check_files_exist_and_exclude(samples_df: pd.DataFrame, path_column: str) -> pd.DataFrame:
+    """
+    Check if files exist at the paths specified in the given column of the DataFrame.
+    Exclude rows where the file does not exist.
+    Inputs:
+        samples_df (pd.DataFrame): The DataFrame containing file paths.
+        path_column (str): The name of the column containing file paths.
+    Outputs:
+        df_filtered (pd.DataFrame): A DataFrame with rows excluded where the file does not exist.
+    """
+    # Check files exist mv this into another function to make DRYer
+    print("Checking files exist...")
+    if samples_df.empty:
+        print("No samples to check.")
+        return samples_df
+    samples_df["file_exists"] = samples_df[path_column].apply(check_file_exists)
+    missing_files_df = samples_df[~samples_df["file_exists"]]
+    print(f"Found {missing_files_df.shape[0]} missing files.")
+    if not missing_files_df.empty:
+        print(missing_files_df[["file_name", f"{path_column}"]])
+    samples_df = samples_df[samples_df["file_exists"]].copy()
+    df_filtered = samples_df.drop(columns=["file_exists"])
+    return df_filtered, missing_files_df
```
changed
<!-- Reviewable comment -ObDeofN81Bos2jWIIFQ:b-n9tkhd -->
<!-- Sent from Reviewable.io -->

RSWilson1 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23)* status: 0 of 6 files reviewed, 5 unresolved discussions (waiting on @rklocke)



<!-- Sent from Reviewable.io -->

RSWilson1 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23)* status: 0 of 6 files reviewed, 6 unresolved discussions (waiting on @rklocke)

___
*[`utils/clarity_extract_handler.py` line 247 at r3](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23#-ObDjKXF8o2b8GB6Ku0M:-ObDjKXF8o2b8GB6Ku0N:b-x2ijzq) ([raw file](https://github.com/eastgenomics/clinvar_submissions/blob/24408948f432dae83e7fb6f1f294488abfa61691/utils/clarity_extract_handler.py#L247)):*
> ```Python
> 
> 
> def filter_duplicate_files(df):
> ```

@rklocke I don't think this makes sense and docstring is inaccurate. 
Trying to work out how to remove files when they aren't the expected normal scenario to allow for manual review.


<!-- Sent from Reviewable.io -->

rklocke reviewed (changes requested):
> @rklocke reviewed 3 of 6 files at r1, 1 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23)* status: 4 of 6 files reviewed, 14 unresolved discussions (waiting on @RSWilson1)

___
*[`tests/test_clarity_extract_handler.py` line 12 at r3](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23#-ObDfrbH3tf2odFo8eRQ:-ObDfrbH3tf2odFo8eRR:bm6f7c9) ([raw file](https://github.com/eastgenomics/clinvar_submissions/blob/24408948f432dae83e7fb6f1f294488abfa61691/tests/test_clarity_extract_handler.py#L12)):*
> ```Python
>     # Copy the provided test_extract.csv to a temp location
>     src = (
>         Path(__file__).parent
> ```

I don't think this file has been added to github?
___
*[`tests/test_clarity_extract_handler.py` line 99 at r3](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23#-ObDcW6l1gX0gB_k2jXt:-ObDcW6l1gX0gB_k2jXu:bcytyp4) ([raw file](https://github.com/eastgenomics/clinvar_submissions/blob/24408948f432dae83e7fb6f1f294488abfa61691/tests/test_clarity_extract_handler.py#L99)):*
> ```Python
>     filtered = ceh.filter_duplicate_files(df)
>     # For sample A, only one file should be kept
>     assert len(filtered[filtered["sample_id"] == "1234567-09877654"]) == 2
> ```

it would be nicer here if rather than multiple asserts you just made a df of what is expected and then used pandas [assert_frame_equal](https://pandas.pydata.org/docs/reference/api/pandas.testing.assert_frame_equal.html)
___
*[`tests/test_clarity_extract_handler.py` line 159 at r3](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23#-ObDe7Xx8eDq9p8T5PB9:-ObDe7Xx8eDq9p8T5PBA:b-cjzb45) ([raw file](https://github.com/eastgenomics/clinvar_submissions/blob/24408948f432dae83e7fb6f1f294488abfa61691/tests/test_clarity_extract_handler.py#L159)):*
> ```Python
>     records = ceh.query_reports_for_project("proj-1", ["123", "456"])
>     assert isinstance(records, list)
>     assert all("sample_id" in r for r in records)
> ```

would it be better to assert what the records should equal? otherwise I'm finding it hard to understand this test
should the result be [], since the sample ID from the filename would be None?
https://github.com/eastgenomics/clinvar_submissions/blob/re-write_new_clarity_extract/utils/clarity_extract_handler.py#L90
___
*[`pandora.py` line 4 at r2](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23#-ObCZjFD8EbstiPbWg-7:-ObCZjFD8EbstiPbWg-8:b-j6k4v3) ([raw file](https://github.com/eastgenomics/clinvar_submissions/blob/35a4dd97ff8bd62140442e71feb9838080e75b29/pandora.py#L4)):*
> ```Python
> Script to add variants from workbooks to Shire database and submit variants
> from Shire database to ClinVar
> Version: 1.0.0
> ```

does this need a version in it? if so does it need to be updated?
___
*[`utils/clarity_extract_handler.py` line 15 at r3](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23#-ObDh24uESritbibE39k:-ObDh24uESritbibE39l:but855f) ([raw file](https://github.com/eastgenomics/clinvar_submissions/blob/24408948f432dae83e7fb6f1f294488abfa61691/utils/clarity_extract_handler.py#L15)):*
> ```Python
> def open_files(clarity):
>     """
>     Open files and read in file contents to DataFrames
> ```

Inputs and outputs in docstring please
___
*[`utils/clarity_extract_handler.py` line 40 at r3](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23#-ObDh4aM6Sf4Syp3Vtki:-ObDh4aM6Sf4Syp3Vtkj:b-l6mwxj) ([raw file](https://github.com/eastgenomics/clinvar_submissions/blob/24408948f432dae83e7fb6f1f294488abfa61691/utils/clarity_extract_handler.py#L40)):*
> ```Python
>     """
>     query_str_for_assays = "|".join(assays)
>     # pattern: str = rf"^002.*_({query_str_for_assays})$"
> ```

remove
___
*[`utils/clarity_extract_handler.py` line 209 at r3](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23#-ObDjF-PBCmlXupEh4tN:-ObDjF-PBCmlXupEh4tO:bqpo9j8) ([raw file](https://github.com/eastgenomics/clinvar_submissions/blob/24408948f432dae83e7fb6f1f294488abfa61691/utils/clarity_extract_handler.py#L209)):*
> ```Python
>     # Clean up run name
>     run_without_prefix = re.sub(r"^002_", "", run)
>     run_without_suffix = re.sub(r"_[^_]*$", "", run_without_prefix)
> ```

does this handle project names with e.g. _38_CEN at the end?
___
*[`utils/clarity_extract_handler.py` line 222 at r3](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23#-ObDjvm0-B3H_vH99WdQ:-ObDjvm0-B3H_vH99WdR:bv050ak) ([raw file](https://github.com/eastgenomics/clinvar_submissions/blob/24408948f432dae83e7fb6f1f294488abfa61691/utils/clarity_extract_handler.py#L222)):*
> ```Python
>         return None
> 
>     return path
> ```

got to here
___
*[`utils/clarity_extract_handler.py` line 235 at r3](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23#-ObDjT_TBaqEAWAaoGCB:-ObDjT_U0LzKYvBGZyOf:bre0ref) ([raw file](https://github.com/eastgenomics/clinvar_submissions/blob/24408948f432dae83e7fb6f1f294488abfa61691/utils/clarity_extract_handler.py#L235)):*
> ```Python
>     """
>     files = list(
>         dxpy.find_data_objects(name=search_query, name_mode="glob", describe=True)
> ```

can just set the describe to return only the name field instead of all fields
___
*[`README.md` line 30 at r3](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23#-ObD_AB3EMhhE43a6IQR:-ObD_AB3EMhhE43a6IQS:b-vfh7mr) ([raw file](https://github.com/eastgenomics/clinvar_submissions/blob/24408948f432dae83e7fb6f1f294488abfa61691/README.md#L30)):*
> ```Markdown
> * `--config`: config file, should be the config.json from https://github.com/eastgenomics/clinvar_submissions_config.
> * `--organisation`: Organisation to submit for, either "CUH" or "NUH". This determines which ClinVar API key and submission URL to use.
> ### One option which specifies the run-mode i.e. from clarity extract or from workbooks:
> ```

from clarity extract, from workbooks path or from samples file?
___
*[`README.md` line 45 at r3](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23#-ObD_vb-4Y3Hvgr3WP4P:-ObD_vb-4Y3Hvgr3WP4Q:bakrbzv) ([raw file](https://github.com/eastgenomics/clinvar_submissions/blob/24408948f432dae83e7fb6f1f294488abfa61691/README.md#L45)):*
> ```Markdown
> 
> **Optional:**
> * `--use_paths`: (boolean) Default is False, if specified as True, will use the paths from the clarity extract directly. If False, will output a CSV for review.
> ```

this and the others below to me are maybe slightly unclear as it seems like from this you would put --use_paths=True rather than just --use_paths or have I got that wrong?


<!-- Sent from Reviewable.io -->

RSWilson1 committed:
test on all PRs

RSWilson1 committed:
updated py in GH actions

RSWilson1 committed:
Removed old function that isn't used.

RSWilson1 committed:
updated dependencies with dxpy

RSWilson1 committed:
updating setup for python

RSWilson1 committed:
Updated gitignore and added test_extract.csv no real data contained.

RSWilson1 committed:
Updated logic to return run name.

RSWilson1 committed:
Changed regex to work for "_genome_build_" run names and combined into single search.

RSWilson1 committed:
rm'd docstring version and TODO

RSWilson1 committed:
Updated README with all examples

RSWilson1 committed:
rm'd comment line

RSWilson1 committed:
Added more accurate regex for run names

RSWilson1 committed:
changed test to use better mock data and fixed to assert list matches expected list.

RSWilson1 committed:
added handling for run name is none and added corresponding tests

RSWilson1 commented on .github/workflows/pytest.yml:
```diff
@@ -5,19 +5,19 @@ on:
     steps:
       - name: Checkout code
-        uses: actions/checkout@v2
+        uses: actions/checkout@v5
```
this is overkill and not needed.
<!-- Reviewable comment -ObSbJntCnivV69AlqIT:bpc3bgu -->
<!-- Sent from Reviewable.io -->

RSWilson1 commented on utils/clarity_extract_handler.py:
```diff
@@ -0,0 +1,383 @@
+        on=["sample_id", "Assay", "report_r_code"],
+        how="inner",
+    ).copy()
+    return report_df_filtered, missing_data_df, multiple_reports_df
```
rm'd
<!-- Reviewable comment -ObS_LRxBDodgFF78dMq:bb7nx99 -->
<!-- Sent from Reviewable.io -->

RSWilson1 commented on utils/clarity_extract_handler.py:
```diff
@@ -0,0 +1,390 @@
+
+    Inputs:
+        assay (str): either CEN or WES/TWE
+        base_path (str or Path): base path to clingen folder
+        filename (str): filename for the xlsx report
+        run (str): sequencing run name
+
+    Outputs:
+        path (Path or None): path to the given filename on clingen
+    """
+    # Handle NaN values
+    if pd.isna(run) or run == "nan":
+        return None
+
+    # Ensure run is a string
+    run = str(run)
+
+    # Ensure base_path is a Path object
+    base_path = Path(base_path)
+
+    # extract run_name from run string
+    run_name_match = re.search(r"(?:002_)?(\d{6}_[A-Z0-9]+_\d{4}_[A-Z0-9]+)", run)
+    run_name = run_name_match.group(1) if run_name_match else None
+
+    # Build path depending on assay
+    if assay == "CEN":
+        path = base_path / assay / "Run folders" / run_name / filename
+    elif assay in ("WES", "TWE"):
+        path = base_path / assay / run_name / filename
+    else:
+        print(
+            f"Warning: Unknown assay '{assay}' for filename {filename}. Returning None."
+        )
+        return None
```
added
<!-- Reviewable comment -ObTGMZI065SeVrlIYTH:b-7tfmjp -->
<!-- Sent from Reviewable.io -->

RSWilson1 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23)* status: 1 of 10 files reviewed, 19 unresolved discussions (waiting on @rklocke)

___
*[`pandora.py` line 4 at r2](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23#-ObCZjFD8EbstiPbWg-7:-ObST6ah4z013SLMEMia:b-vf3nrc) ([raw file](https://github.com/eastgenomics/clinvar_submissions/blob/35a4dd97ff8bd62140442e71feb9838080e75b29/pandora.py#L4)):*
<details><summary><i>Previously, rklocke (Becky Locke) wrote…</i></summary><blockquote>

does this need a version in it? if so does it need to be updated?
</blockquote></details>

I agree this is left over from previous release as the version is handled in GH, this isn't needed. rm'd
___
*[`README.md` line 30 at r3](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23#-ObD_AB3EMhhE43a6IQR:-ObTGOLp-SMTz6OAphs0:b-7tfmjp) ([raw file](https://github.com/eastgenomics/clinvar_submissions/blob/24408948f432dae83e7fb6f1f294488abfa61691/README.md#L30)):*
<details><summary><i>Previously, rklocke (Becky Locke) wrote…</i></summary><blockquote>

from clarity extract, from workbooks path or from samples file?
</blockquote></details>

added
___
*[`README.md` line 45 at r3](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23#-ObD_vb-4Y3Hvgr3WP4P:-ObSTGva6xBlZtSIhoS-:bxtwc5w) ([raw file](https://github.com/eastgenomics/clinvar_submissions/blob/24408948f432dae83e7fb6f1f294488abfa61691/README.md#L45)):*
<details><summary><i>Previously, rklocke (Becky Locke) wrote…</i></summary><blockquote>

this and the others below to me are maybe slightly unclear as it seems like from this you would put --use_paths=True rather than just --use_paths or have I got that wrong?
</blockquote></details>

--use_paths when given in cmd line will lead to argparse storing the boolean so it's not run as --use_paths=True.
___
*[`tests/test_clarity_extract_handler.py` line 12 at r3](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23#-ObDfrbH3tf2odFo8eRQ:-ObSTaSl5EUo_8bHgTCV:b-6978hb) ([raw file](https://github.com/eastgenomics/clinvar_submissions/blob/24408948f432dae83e7fb6f1f294488abfa61691/tests/test_clarity_extract_handler.py#L12)):*
<details><summary><i>Previously, rklocke (Becky Locke) wrote…</i></summary><blockquote>

I don't think this file has been added to github?
</blockquote></details>

added now
___
*[`tests/test_clarity_extract_handler.py` line 99 at r3](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23#-ObDcW6l1gX0gB_k2jXt:-ObSd_ZU14OqSl0XlyEm:b-5qyn15) ([raw file](https://github.com/eastgenomics/clinvar_submissions/blob/24408948f432dae83e7fb6f1f294488abfa61691/tests/test_clarity_extract_handler.py#L99)):*
<details><summary><i>Previously, rklocke (Becky Locke) wrote…</i></summary><blockquote>

it would be nicer here if rather than multiple asserts you just made a df of what is expected and then used pandas [assert_frame_equal](https://pandas.pydata.org/docs/reference/api/pandas.testing.assert_frame_equal.html)
</blockquote></details>

not sure this was the correct place for the comment but made changes to the other test where this makes sense?
___
*[`tests/test_clarity_extract_handler.py` line 159 at r3](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23#-ObDe7Xx8eDq9p8T5PB9:-ObTGQCz8f087sWAjZLK:b-n9tkhd) ([raw file](https://github.com/eastgenomics/clinvar_submissions/blob/24408948f432dae83e7fb6f1f294488abfa61691/tests/test_clarity_extract_handler.py#L159)):*
<details><summary><i>Previously, rklocke (Becky Locke) wrote…</i></summary><blockquote>

would it be better to assert what the records should equal? otherwise I'm finding it hard to understand this test
should the result be [], since the sample ID from the filename would be None?
https://github.com/eastgenomics/clinvar_submissions/blob/re-write_new_clarity_extract/utils/clarity_extract_handler.py#L90
</blockquote></details>

changed
___
*[`utils/clarity_extract_handler.py` line 15 at r3](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23#-ObDh24uESritbibE39k:-ObSSlsd2dnjCAA4yfK4:b-7tfmjp) ([raw file](https://github.com/eastgenomics/clinvar_submissions/blob/24408948f432dae83e7fb6f1f294488abfa61691/utils/clarity_extract_handler.py#L15)):*
<details><summary><i>Previously, rklocke (Becky Locke) wrote…</i></summary><blockquote>

Inputs and outputs in docstring please
</blockquote></details>

added
___
*[`utils/clarity_extract_handler.py` line 40 at r3](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23#-ObDh4aM6Sf4Syp3Vtki:-ObSSJvo3TiAuplYRbva:bb7nx99) ([raw file](https://github.com/eastgenomics/clinvar_submissions/blob/24408948f432dae83e7fb6f1f294488abfa61691/utils/clarity_extract_handler.py#L40)):*
<details><summary><i>Previously, rklocke (Becky Locke) wrote…</i></summary><blockquote>

remove
</blockquote></details>

rm'd
___
*[`utils/clarity_extract_handler.py` line 209 at r3](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23#-ObDjF-PBCmlXupEh4tN:-ObSTiBz1LE-JIFy3DKO:b-w2ghji) ([raw file](https://github.com/eastgenomics/clinvar_submissions/blob/24408948f432dae83e7fb6f1f294488abfa61691/utils/clarity_extract_handler.py#L209)):*
<details><summary><i>Previously, rklocke (Becky Locke) wrote…</i></summary><blockquote>

does this handle project names with e.g. _38_CEN at the end?
</blockquote></details>

updated to handle this, need to test.
___
*[`utils/clarity_extract_handler.py` line 222 at r3](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23#-ObDjvm0-B3H_vH99WdQ:-ObSSo7qD8y1rTM71Eir:b-896fix) ([raw file](https://github.com/eastgenomics/clinvar_submissions/blob/24408948f432dae83e7fb6f1f294488abfa61691/utils/clarity_extract_handler.py#L222)):*
<details><summary><i>Previously, rklocke (Becky Locke) wrote…</i></summary><blockquote>

got to here
</blockquote></details>

Done.
___
*[`utils/clarity_extract_handler.py` line 235 at r3](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23#-ObDjT_TBaqEAWAaoGCB:-ObSSpUt9O7IZNfEf7XQ:b-n9tkhd) ([raw file](https://github.com/eastgenomics/clinvar_submissions/blob/24408948f432dae83e7fb6f1f294488abfa61691/utils/clarity_extract_handler.py#L235)):*
<details><summary><i>Previously, rklocke (Becky Locke) wrote…</i></summary><blockquote>

can just set the describe to return only the name field instead of all fields
</blockquote></details>

changed
___
*[`utils/clarity_extract_handler.py` line 247 at r3](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23#-ObDjKXF8o2b8GB6Ku0M:-ObDjrCi8iztHlg0_cVL:b-n3tao8) ([raw file](https://github.com/eastgenomics/clinvar_submissions/blob/24408948f432dae83e7fb6f1f294488abfa61691/utils/clarity_extract_handler.py#L247)):*
<details><summary><i>Previously, RSWilson1 (Robert Wilson) wrote…</i></summary><blockquote>

@rklocke I don't think this makes sense and docstring is inaccurate. 
Trying to work out how to remove files when they aren't the expected normal scenario to allow for manual review.
</blockquote></details>

any thoughts welcome :)  I guess the normal expected scenario 


<!-- Sent from Reviewable.io -->

RSWilson1 committed:
changed output_inconsistent_files function to use None as defaults and handling for None including timestamp.

RSWilson1 committed:
Output inconsistent files when --use_paths not specified for manual review.

RSWilson1 committed:
fixed multiple_reports_df to default to empty df.

RSWilson1 committed:
rm'd debug print

RSWilson1 commented on pandora.py:
```diff
@@ -1,101 +1,142 @@
+            f"clarity extract. See duplicate_data_clarity_extract_{timestamp}.csv for details."
+        )
+        duplicate_data_df.to_csv(f"duplicate_data_clarity_extract_{timestamp}.csv", index=False)
+
```
changed
<!-- Reviewable comment -ObWqI6Y6YvSoBVcGlg4:b-n9tkhd -->
<!-- Sent from Reviewable.io -->

RSWilson1 commented on utils/clarity_extract_handler.py:
```diff
@@ -0,0 +1,390 @@
+        print(f"Extracted assay from {filename}: {match.group(1) if match else 'None'}")
+        return match.group(1) if match else None
+
+    report_df["Assay"] = report_df["file_name"].apply(extract_assay_from_filename)
```
rm'd
<!-- Reviewable comment -ObWr6ll7wX51ko_xmhQ:bb7nx99 -->
<!-- Sent from Reviewable.io -->

RSWilson1 commented on utils/clarity_extract_handler.py:
```diff
@@ -0,0 +1,390 @@
+            how="inner",
+        ).copy()
+    else:
+        print("No samples with multiple reports per assay found.")
```
changed
<!-- Reviewable comment -ObWqwI75kel2pnA3qDz:b-n9tkhd -->
<!-- Sent from Reviewable.io -->

RSWilson1 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23)* status: 1 of 10 files reviewed, 19 unresolved discussions (waiting on @rklocke)



<!-- Sent from Reviewable.io -->

RSWilson1 committed:
Updated workflow diagram

RSWilson1 committed:
typo

RSWilson1 commented on README.md:
```diff
@@ -26,14 +26,29 @@ This script searches a given folder for Excel variant workbooks that have not pr
-* `--samples_file`: Local path to a CSV file containing sample information. If specified, the script will use this file instead of searching for workbooks. The CSV should have columns for `file_name`, `sample_id`, `workbook_path`.
-* `--dry-run`: (boolean) Default is False, if specified as True, will not make any changes to the database or submit to ClinVar. Useful for testing.
+* `--dry_run`: (boolean) Default is False, if specified as True, will not make any changes to the database or submit to ClinVar. Useful for testing.
+
```
typo fixed
<!-- Reviewable comment -ObXMwb529hJgGqeWFEE:b4ax0f1 -->
<!-- Sent from Reviewable.io -->

RSWilson1 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23)* status: 1 of 12 files reviewed, 19 unresolved discussions (waiting on @rklocke)



<!-- Sent from Reviewable.io -->

rklocke reviewed (changes requested):
> @rklocke reviewed 1 of 2 files at r3, 1 of 3 files at r5, 1 of 1 files at r6, 2 of 5 files at r7, 3 of 3 files at r9, 3 of 3 files at r10.
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23)* status: 11 of 12 files reviewed, 4 unresolved discussions (waiting on @RSWilson1)

___
*[`README.md` line 45 at r3](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23#-ObD_vb-4Y3Hvgr3WP4P:-ObXC5DODvlS_Q8_GkPO:bnzdmj2) ([raw file](https://github.com/eastgenomics/clinvar_submissions/blob/24408948f432dae83e7fb6f1f294488abfa61691/README.md#L45)):*
<details><summary><i>Previously, RSWilson1 (Robert Wilson) wrote…</i></summary><blockquote>

--use_paths when given in cmd line will lead to argparse storing the boolean so it's not run as --use_paths=True.
</blockquote></details>

I know but this says "if specified as True", wouldn't it make more sense to someone reading this to just say "if specified"?
___
*[`tests/test_clarity_extract_handler.py` line 66 at r10](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23#-ObXEENA2y1oYa7eh3ra:-ObXEENA2y1oYa7eh3rb:b7e8dxb) ([raw file](https://github.com/eastgenomics/clinvar_submissions/blob/7475cdcf44ce8dd703e1a3dc3ff1d8702fd43599/tests/test_clarity_extract_handler.py#L66)):*
> ```Python
>     assert str(exc.value).startswith("Error reading clarity extract:")
> 
> #002_251010_A01303_0320_BACWV9DRX7_37_CEN
> ```

are these meant to be here? or can some context be given if they're staying?
___
*[`utils/clarity_extract_handler.py` line 324 at r10](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23#-ObXF03T1PxaQr2WzgIP:-ObXF03T1PxaQr2WzgIQ:b-2do7z2) ([raw file](https://github.com/eastgenomics/clinvar_submissions/blob/7475cdcf44ce8dd703e1a3dc3ff1d8702fd43599/utils/clarity_extract_handler.py#L324)):*
> ```Python
> 
>     # Check if report_r_code is in R_codes list
>     def is_report_code_in_list(row):
> ```

I'm not sure about these functions within a function? Can we separate them out and give them proper docstrings etc?

There is also so much going on in this one function - reading, extracting R codes and assays, getting reports, filtering - please can they be split out so this is easier to understand / test /maintain?

I was trying to test this out to see what it's doing using the test clarity CSV, if nothing is found in dnanexus I feel like a lot is done to an empty df afterwards


<!-- Sent from Reviewable.io -->

RSWilson1 committed:
rm'd comments and formatted.

RSWilson1 committed:
clearer language about cmd line args.

RSWilson1 committed:
Adding dnanexus authentication function

RSWilson1 committed:
fix sqlalchemy requirement

RSWilson1 committed:
Split the handle_clarity_extract function into component parts and moved nested functions to normal functions.
Changed regex to work with or without genome_build in the run name.
Changed assay search to reflect files as TWE isn't present.
Combined filtering CNV/mosaics to single cmd

RSWilson1 committed:
Fixed R code checking and added more tests.

RSWilson1 committed:
Added output_dir input to allow control of where to output files.

RSWilson1 committed:
Update readme for new input

RSWilson1 committed:
raise value error for wrong type in R_codes.

RSWilson1 committed:
example data for clarity extract for checking R codes

RSWilson1 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23)* status: 6 of 15 files reviewed, 8 unresolved discussions (waiting on @rklocke)

___
*[`README.md` line 45 at r3](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23#-ObD_vb-4Y3Hvgr3WP4P:-ObXoGf827wBfpBtpI_4:b-ul0ixl) ([raw file](https://github.com/eastgenomics/clinvar_submissions/blob/24408948f432dae83e7fb6f1f294488abfa61691/README.md#L45)):*
<details><summary><i>Previously, rklocke (Becky Locke) wrote…</i></summary><blockquote>

I know but this says "if specified as True", wouldn't it make more sense to someone reading this to just say "if specified"?
</blockquote></details>

Oh I see what you mean now! Yep changed.
___
*[`tests/test_clarity_extract_handler.py` line 66 at r10](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23#-ObXEENA2y1oYa7eh3ra:-ObXoM170Th0IvgPVnLQ:bb7nx99) ([raw file](https://github.com/eastgenomics/clinvar_submissions/blob/7475cdcf44ce8dd703e1a3dc3ff1d8702fd43599/tests/test_clarity_extract_handler.py#L66)):*
<details><summary><i>Previously, rklocke (Becky Locke) wrote…</i></summary><blockquote>

are these meant to be here? or can some context be given if they're staying?
</blockquote></details>

rm'd
___
*[`utils/clarity_extract_handler.py` line 324 at r10](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23#-ObXF03T1PxaQr2WzgIP:-ObY2WUE2YARCzc11uDm:by4qmn9) ([raw file](https://github.com/eastgenomics/clinvar_submissions/blob/7475cdcf44ce8dd703e1a3dc3ff1d8702fd43599/utils/clarity_extract_handler.py#L324)):*
<details><summary><i>Previously, rklocke (Becky Locke) wrote…</i></summary><blockquote>

I'm not sure about these functions within a function? Can we separate them out and give them proper docstrings etc?

There is also so much going on in this one function - reading, extracting R codes and assays, getting reports, filtering - please can they be split out so this is easier to understand / test /maintain?

I was trying to test this out to see what it's doing using the test clarity CSV, if nothing is found in dnanexus I feel like a lot is done to an empty df afterwards
</blockquote></details>

Split into multiple functions and moved out the nested functions.


<!-- Sent from Reviewable.io -->

RSWilson1 committed:
rm'd redundant check

RSWilson1 committed:
Fixed naming conventions for create_path

RSWilson1 committed:
rm'd debug csv outputs

RSWilson1 committed:
fixed valueerror to typeerror

RSWilson1 commented on utils/clarity_extract_handler.py:
```diff
@@ -0,0 +1,541 @@
+
+def filtering_reports(report_df):
+    """
+    Filter out reports based on if filename doesn't match the verified R codes
+    and if filename contains _CNV_ or _mosaic_
+
+    Inputs:
+        report_df (pd.DataFrame):
+            DataFrame containing report information with a 'file_name' column
+            and 'R_codes' column.
+
+    Outputs:
+        filtered_df (pd.DataFrame): Filtered DataFrame excluding rows with '_CNV_' or '_mosaic_' in 'file_name'.
+    """
+    if report_df.empty or report_df is None:
+        print("Report DataFrame is empty or None. Skipping filtering.")
+        return report_df
+
+    report_df.to_csv("test_all_reports_df.csv", index=False)
+    report_df["rcode_match"] = report_df.apply(is_report_code_in_list, axis=1)
+    # Filter rows where R code doesn't match
+    report_df = report_df[report_df["rcode_match"]].copy()
+    # clean up by dropping the rcode_match column
+    report_df = report_df.drop(columns=["rcode_match"])
+    # Filter out rows where filename contains _CNV_ or _mosaic_
+    filtered_df = report_df[
+        ~report_df["file_name"].str.contains(r"_(CNV|mosaic)_", na=False)
+    ]
+    filtered_df.to_csv("test_filtered_reports.csv", index=False)
+
+    return filtered_df
```
rm'd
<!-- Reviewable comment -ObhIqDvAPhBmDTTCR9D:bb7nx99 -->
<!-- Sent from Reviewable.io -->

RSWilson1 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23)* status: 6 of 15 files reviewed, 9 unresolved discussions (waiting on @rklocke)



<!-- Sent from Reviewable.io -->

RSWilson1 committed:
Fixed tests
Fixed default for fetch_all_reports
Added more error handling and propagate errors using 'from e'

RSWilson1 committed:
Changed imports for csvs for tests.
Updated error types
changed check order for report_df
added additional check for filename
added coverage to pytest GH action

RSWilson1 committed:
update requirements for pytest-cov

RSWilson1 committed:
Updated to include toml file

RSWilson1 commented on tests/test_clarity_extract_handler.py:
```diff
@@ -0,0 +1,307 @@
+    def example_df(self):
+        return pd.read_csv(
+            "tests/test_data/clarity_extract_examples/test_all_reports_df.csv"
+        )
```
Not needed as works for testing
<!-- Reviewable comment -Obhc4HlAhecAFZRUZea:b-599t22 -->
<!-- Sent from Reviewable.io -->

RSWilson1 commented on utils/clarity_extract_handler.py:
```diff
@@ -0,0 +1,541 @@
+    except pd.errors.EmptyDataError:
+        raise ValueError(f"Clarity extract file is empty: {clarity}")
+    except Exception as e:
+        raise Exception(f"Error reading clarity extract: {e}")
```
added
<!-- Reviewable comment -ObhcHzUBid8shcN0BHE:b-7tfmjp -->
<!-- Sent from Reviewable.io -->

RSWilson1 commented on utils/clarity_extract_handler.py:
```diff
@@ -0,0 +1,543 @@
+        set number of ids to process at a time, by default 100
+    max_workers : int, optional
+        number of maximum workers to destribute across, by default 16
+
```
done
<!-- Reviewable comment -ObhcJVe9TW6zMwvQRu1:bb7f2mf -->
<!-- Sent from Reviewable.io -->

RSWilson1 commented on utils/clarity_extract_handler.py:
```diff
@@ -0,0 +1,543 @@
+    return target in normalized_codes
+
+
+def extract_assay_from_filename(filename):
+    """
+    Extract the assay type from the given filename.
+
+    Parameters
+    ----------
+    filename : str
+        The name of the file from which to extract the assay type.
+
+    Returns
+    -------
+    str or None
+        The extracted assay type (CEN or WES) or None if not found.
+    """
+    if pd.isna(filename):
+        return None
+    match = re.search(r"(CEN|WES)", filename)
+    return match.group(1) if match else None
+
```
Files don't have TWE so not needed
<!-- Reviewable comment -ObhbcPu-kSrKVQfpTlI:b-8xky5u -->
<!-- Sent from Reviewable.io -->

RSWilson1 commented on utils/clarity_extract_handler.py:
```diff
@@ -0,0 +1,543 @@
+    if report_df.empty or report_df is None:
+        print("Report DataFrame is empty or None. Skipping filtering.")
+        return report_df
+
```
done
<!-- Reviewable comment -Obhby61C9VSj6xa9xgb:bb7f2mf -->
<!-- Sent from Reviewable.io -->

RSWilson1 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23)* status: 5 of 16 files reviewed, 8 unresolved discussions (waiting on @rklocke)



<!-- Sent from Reviewable.io -->

RSWilson1 committed:
rm'd redudnant check

RSWilson1 committed:
print to see the submission response

RSWilson1 committed:
Gate dnanexus authentication for only when needed.

RSWilson1 commented on pandora.py:
```diff
@@ -104,9 +188,11 @@ def main():
     cuh_header = clinvar.create_header(cuh_api_key)
     nuh_header = clinvar.create_header(nuh_api_key)
+    utils.dx_login(dnanexus_api_key)
     api_url = utils.select_api_url(args.clinvar_testing, config)
```
wrong addressed commit but yes fixed
<!-- Reviewable comment -OblN8HbEclxxHn6TcsK:b-d17rd4 -->
<!-- Sent from Reviewable.io -->

RSWilson1 commented on pyproject.toml:
```diff
@@ -0,0 +1,22 @@
+[project]
+name = "clinvar-submissions"
+version = "1.1.0"
+description = "Submit variants to ClinVar from Excel workbooks"
+readme = "README.md"
+requires-python = ">=3.12"
```
not needed
<!-- Reviewable comment -OblODXr0R7rsy17xdRe:bpp6stj -->
<!-- Sent from Reviewable.io -->

RSWilson1 commented on pyproject.toml:
```diff
@@ -0,0 +1,22 @@
+description = "Submit variants to ClinVar from Excel workbooks"
+readme = "README.md"
+requires-python = ">=3.12"
+dependencies = [
+    "freezegun==1.5.1",
+    "numpy==1.26.4",
+    "openpyxl==3.1.2",
+    "pandas==2.2.0",
+    "psycopg2-binary==2.9.10",
+    "pytest==8.0.0",
+    "requests==2.32.4",
+    "sqlalchemy==2.0.0",
+    "coverage==6.5.0",
+    "pytest-cov==4.0.0",
+    "dxpy>=0.397.0",
+    "pip>=25.2",
+    "mypy>=1.17.1",
+    "datetime>=5.5",
+]
```
changed
<!-- Reviewable comment -OblOIUBDMl4CEFZXmX6:b-n9tkhd -->
<!-- Sent from Reviewable.io -->

RSWilson1 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23)* status: 5 of 16 files reviewed, 9 unresolved discussions (waiting on @rklocke)



<!-- Sent from Reviewable.io -->

RSWilson1 committed:
added example api key to json in README

RSWilson1 committed:
Updated print and docstring

RSWilson1 commented on pandora.py:
```diff
@@ -1,100 +1,184 @@
+    Output any inconsistent files from clarity extract handling for review
+    Inputs:
+        missing_data_df (pd.DataFrame): dataframe of samples with missing data
+        duplicate_data_df (pd.DataFrame): dataframe of samples with duplicate data
+        timestamp (str): timestamp to append to filenames
+    Outputs:
+        None
+    Side effects:
+        Outputs CSV files if any inconsistent data found
+    """
+    if missing_data_df is None:
+        missing_data_df = pd.DataFrame()
+    if duplicate_data_df is None:
+        duplicate_data_df = pd.DataFrame()
+    if timestamp is None:
+        print("Timestamp not provided. Not outputting inconsistent files.")
+        return
+
+    # Output any inconsistent files for review
+    if not missing_data_df.empty:
+        print(
+            f"{missing_data_df.shape[0]} samples with missing data found in "
+            f"clarity extract. See missing_data_clarity_extract_{timestamp}.csv for details."
+        )
+        path_to_missing = os.path.join(
+            output_dir, f"missing_data_clarity_extract_{timestamp}.csv"
+        )
+        missing_data_df.to_csv(path_to_missing, index=False)
+    if not duplicate_data_df.empty:
+        print(
+            f"{duplicate_data_df.shape[0]} duplicate samples found in "
+            f"clarity extract. See duplicate_data_clarity_extract_{timestamp}.csv for details."
+        )
+        path_to_duplicate = os.path.join(
+            output_dir, f"duplicate_data_clarity_extract_{timestamp}.csv"
+        )
+        duplicate_data_df.to_csv(path_to_duplicate, index=False)
+    if missing_data_df.empty and duplicate_data_df.empty:
+        print("No inconsistent data found in clarity extract.")
+
```
changed
<!-- Reviewable comment -Oblue2eErH0mfNGMHtC:b-n9tkhd -->
<!-- Sent from Reviewable.io -->

RSWilson1 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23)* status: 5 of 16 files reviewed, 4 unresolved discussions (waiting on @rklocke)



<!-- Sent from Reviewable.io -->

rklocke reviewed (approved):
> @rklocke reviewed 2 of 6 files at r12, 4 of 6 files at r13, 5 of 5 files at r16, 1 of 1 files at r18, 1 of 1 files at r19, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/clinvar_submissions/23)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @RSWilson1)



<!-- Sent from Reviewable.io -->
